Bug: no-invalid-this: false positive with this in class field

### Environment

Node version: v16.13.1
npm version: 8.3.0
Local ESLint version: 8.6.0
Global ESLint version: N/A
Operating System: Ubuntu 20.04.2 LTS


### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```JSON
{
    "parserOptions": {
        "ecmaVersion": "latest",
        "sourceType": "module"
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```
</details>

```js
class Foo {
    bar = this;
}
```


### What did you expect to happen?

No error reported.

### What actually happened?

```
$ npx eslint index.js 

/home/regseb/testcase/index.js
  2:11  error  Unexpected 'this'  no-invalid-this

✖ 1 problem (1 error, 0 warnings)
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

No error reported with this configuration (without `"sourceType": "module"`):

```JSON
{
    "parserOptions": {
        "ecmaVersion": "latest"
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```

--- a/lib/rules/no-invalid-this.js
+++ b/lib/rules/no-invalid-this.js
@@ -20,7 +20,7 @@ const astUtils = require("./utils/ast-utils");
  * That is, if `this` within the code path refers to `this` of surrounding code path.
  * @param {CodePath} codePath Code path.
  * @param {ASTNode} node Node that started the code path.
- * @returns {void}
+ * @returns {boolean} `true` if it is a code path with lexical `this` binding.
  */
 function isCodePathWithLexicalThis(codePath, node) {
     return codePath.origin === "function" && node.type === "ArrowFunctionExpression";

--- a/lib/rules/no-invalid-this.js
+++ b/lib/rules/no-invalid-this.js
@@ -12,6 +12,21 @@
 const astUtils = require("./utils/ast-utils");
 
 //------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Determines if the given code path is a code path with lexical `this` binding.
+ * That is, if `this` within the code path refers to `this` of surrounding code path.
+ * @param {CodePath} codePath Code path.
+ * @param {ASTNode} node Node that started the code path.
+ * @returns {void}
+ */
+function isCodePathWithLexicalThis(codePath, node) {
+    return codePath.origin === "function" && node.type === "ArrowFunctionExpression";
+}
+
+//------------------------------------------------------------------------------
 // Rule Definition
 //------------------------------------------------------------------------------
 
@@ -72,70 +87,52 @@ module.exports = {
             return current;
         };
 
-        /**
-         * Pushs new checking context into the stack.
-         *
-         * The checking context is not initialized yet.
-         * Because most functions don't have `this` keyword.
-         * When `this` keyword was found, the checking context is initialized.
-         * @param {ASTNode} node A function node that was entered.
-         * @returns {void}
-         */
-        function enterFunction(node) {
-
-            // `this` can be invalid only under strict mode.
-            stack.push({
-                init: !context.getScope().isStrict,
-                node,
-                valid: true
-            });
-        }
-
-        /**
-         * Pops the current checking context from the stack.
-         * @returns {void}
-         */
-        function exitFunction() {
-            stack.pop();
-        }
-
         return {
 
-            /*
-             * `this` is invalid only under strict mode.
-             * Modules is always strict mode.
-             */
-            Program(node) {
-                const scope = context.getScope(),
-                    features = context.parserOptions.ecmaFeatures || {};
+            onCodePathStart(codePath, node) {
+                if (isCodePathWithLexicalThis(codePath, node)) {
+                    return;
+                }
 
+                if (codePath.origin === "program") {
+                    const scope = context.getScope();
+                    const features = context.parserOptions.ecmaFeatures || {};
+
+                    stack.push({
+                        init: true,
+                        node,
+                        valid: !(
+                            scope.isStrict ||
+                            node.sourceType === "module" ||
+                            (features.globalReturn && scope.childScopes[0].isStrict)
+                        )
+                    });
+
+                    return;
+                }
+
+                /*
+                 * `init: false` means that `valid` isn't determined yet.
+                 * Most functions don't use `this`, and the calculation for `valid`
+                 * is relatively costly, so we'll calculate it lazily when the first
+                 * `this` within the function is traversed. A special case are non-strict
+                 * functions, because `this` refers to the global object and therefore is
+                 * always valid, so we can set `init: true` right away.
+                 */
                 stack.push({
-                    init: true,
+                    init: !context.getScope().isStrict,
                     node,
-                    valid: !(
-                        scope.isStrict ||
-                        node.sourceType === "module" ||
-                        (features.globalReturn && scope.childScopes[0].isStrict)
-                    )
+                    valid: true
                 });
             },
 
-            "Program:exit"() {
+            onCodePathEnd(codePath, node) {
+                if (isCodePathWithLexicalThis(codePath, node)) {
+                    return;
+                }
+
                 stack.pop();
             },
-
-            FunctionDeclaration: enterFunction,
-            "FunctionDeclaration:exit": exitFunction,
-            FunctionExpression: enterFunction,
-            "FunctionExpression:exit": exitFunction,
-
-            // Field initializers are implicit functions.
-            "PropertyDefinition > *.value": enterFunction,
-            "PropertyDefinition > *.value:exit": exitFunction,
-
-            // Class static blocks are implicit functions.
-            StaticBlock: enterFunction,
-            "StaticBlock:exit": exitFunction,
 
             // Reports if `this` of the current context is invalid.
             ThisExpression(node) {

--- a/tests/lib/rules/no-invalid-this.js
+++ b/tests/lib/rules/no-invalid-this.js
@@ -119,6 +119,15 @@ const patterns = [
         valid: [NORMAL],
         invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
+    {
+        code: "() => { this }; this;",
+        parserOptions: {
+            ecmaVersion: 6
+        },
+        errors,
+        valid: [NORMAL],
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
+    },
 
     // IIFE.
     {
@@ -746,6 +755,18 @@ const patterns = [
 
     // Class fields.
     {
+        code: "class C { field = this }",
+        parserOptions: { ecmaVersion: 2022 },
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+        invalid: []
+    },
+    {
+        code: "class C { static field = this }",
+        parserOptions: { ecmaVersion: 2022 },
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+        invalid: []
+    },
+    {
         code: "class C { field = console.log(this); }",
         parserOptions: { ecmaVersion: 2022 },
         valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
@@ -774,6 +795,20 @@ const patterns = [
         parserOptions: { ecmaVersion: 2022 },
         valid: [NORMAL], // the global this in non-strict mode is OK.
         invalid: [USE_STRICT, IMPLIED_STRICT, MODULES],
+        errors: [{ messageId: "unexpectedThis", type: "ThisExpression" }]
+    },
+    {
+        code: "class C { foo = () => this; }",
+        parserOptions: { ecmaVersion: 2022 },
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+        invalid: [],
+        errors: [{ messageId: "unexpectedThis", type: "ThisExpression" }]
+    },
+    {
+        code: "class C { foo = () => { this }; }",
+        parserOptions: { ecmaVersion: 2022 },
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+        invalid: [],
         errors: [{ messageId: "unexpectedThis", type: "ThisExpression" }]
     },
 
@@ -815,6 +850,13 @@ const patterns = [
         parserOptions: { ecmaVersion: 2022 },
         valid: [],
         invalid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+        errors: [{ messageId: "unexpectedThis", type: "ThisExpression" }]
+    },
+    {
+        code: "class C { static {} [this]; }",
+        parserOptions: { ecmaVersion: 2022 },
+        valid: [NORMAL],
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES],
         errors: [{ messageId: "unexpectedThis", type: "ThisExpression" }]
     },
     {

